hibernate-core/src/main/java/org/hibernate/sql/results/spi/ListResultsConsumer.java

```diff
@@ -68,6 +74,14 @@ public ListResultsConsumer(boolean uniqueFilter) {
 				boolean add = true;
 				if ( uniqueRows ) {
 					if ( results.contains( row ) ) {
```

Comment:
It's not part of your PR, but I guess it might be more appropriate to use the backing `JavaTypeDescriptor` and do the unique check based on that. I also think that we should create another loop for the uniquig part. Something like the following:

```
if ( uniqueSemantics == NONE ) {
	while ( rowProcessingState.next() ) {
		results.add( rowReader.readRow( rowProcessingState, processingOptions ) );
		rowProcessingState.finishRowProcessing();
	}
}
else {
	final JavaTypeDescriptor<R> jtd = rowReader.getResultJavaTypeDescriptor();
	while ( rowProcessingState.next() ) {
		final R row = rowReader.readRow( rowProcessingState, processingOptions );
		boolean add = true;
		for ( R existingRow : results ) {
			if ( jtd.isEqual( existingRow, row, session ) ) {
				if ( singleResultExpected && !rowProcessingState.hasCollectionInitializers() ) {
					throw new HibernateException(
							"More than one row with the given identifier was found: " +
								jdbcValuesSourceProcessingState.getExecutionContext().getEntityId() +
								", for class: " +
								resultJavaType.getName()
					);
				}
				add = false;
				break;
			}
		}
		if ( add ) {
			results.add( row );
		}
		rowProcessingState.finishRowProcessing();
	}
}
```

Wdyt?

Reply:
I'll try it

Reply:
@beikov not sure how to get the result `JavaTypeDescriptor` from the `RowReader` in a clean way, any suggestion?

Reply:
You would have to expose it in `RowReader`.